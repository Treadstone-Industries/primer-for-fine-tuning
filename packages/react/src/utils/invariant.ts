function emptyFunction() {}

// Inspired by invariant by fbjs
// @see https://github.com/facebook/fbjs/blob/main/packages/fbjs/src/__forks__/invariant.js
const invariant = __DEV__
  ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function invariant(condition: any, format?: string, ...args: Array<any>) {
      if (!condition) {
        let error

        if (format === undefined) {
          error = new Error(
            'Minified exception occurred; use the non-minified dev environment ' +
              'for the full error message and additional helpful warnings.',
          )
        } else {
          let index = 0
          const message = format.replace(/%s/g, () => {
            return args[index++]
          })

          error = new Error(message)
          error.name = 'Invariant Violation'
        }

        throw error
      }
    }
  : emptyFunction

export {invariant}
