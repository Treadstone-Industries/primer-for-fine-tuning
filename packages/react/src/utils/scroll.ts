/**
 * Returns the nearest scrollable parent of the element or `null` if the element
 * is not contained in a scrollable element.
 */
export function getScrollContainer(element: Element | null): Element | null {
  if (!element || element === document.body) {
    return null
  }

  return isScrollable(element) ? element : getScrollContainer(element.parentElement)
}

/** Returns `true` if the element is scrollable */
function isScrollable(element: Element) {
  const hasScrollableContent = element.scrollHeight > element.clientHeight

  return hasScrollableContent && !(window.getComputedStyle(element).overflowY.indexOf('hidden') !== -1)
}
