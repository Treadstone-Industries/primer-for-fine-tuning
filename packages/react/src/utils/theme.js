// Utility functions used in theme-preval.js
// This file needs to be a JavaScript file using CommonJS to be compatible with preval

const isEmpty = require('lodash.isempty')
const isObject = require('lodash.isobject')

function fontStack(fonts) {
  return fonts.map(font => (font.includes(' ') ? `"${font}"` : font)).join(', ')
}

// The following functions are a temporary measure for splitting shadow values out from the colors object.
// Eventually, we will push these structural changes upstream to primer/primitives so this data manipulation
// will not be needed.

function isShadowValue(value) {
  return typeof value === 'string' && /(inset\s|)([0-9.]+(\w*)\s){1,4}(rgb[a]?\(.*\)|\w+)/.test(value)
}

function isColorValue(value) {
  if (isShadowValue(value)) return false
  if (value.startsWith('#')) return true // #hex
  if (value.startsWith('rgb')) return true // rgb, rgba
  if (value.startsWith('hsl')) return true // hsl, hsla
  if (value.startsWith('var')) return true // var(--color)
  return false
}

function filterObject(obj, predicate) {
  if (Array.isArray(obj)) {
    return obj.filter(predicate)
  }

  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (isObject(value)) {
      const result = filterObject(value, predicate)

      // Don't include empty objects or arrays
      if (!isEmpty(result)) {
        acc[key] = result
      }
    } else if (predicate(value)) {
      acc[key] = value
    }

    return acc
  }, {})
}

function partitionColors(colors) {
  return {
    colors: filterObject(colors, value => isColorValue(value)),
    shadows: filterObject(colors, value => isShadowValue(value)),
  }
}

function omitScale(obj) {
  const {scale, ...rest} = obj
  return rest
}

module.exports = {
  fontStack,
  isShadowValue,
  isColorValue,
  filterObject,
  partitionColors,
  omitScale,
}
