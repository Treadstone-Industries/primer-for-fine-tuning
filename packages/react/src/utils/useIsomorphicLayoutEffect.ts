// eslint-disable-next-line no-restricted-imports
import {useEffect, useLayoutEffect} from 'react'

const useIsomorphicLayoutEffect =
  typeof window !== 'undefined' &&
  // eslint-disable-next-line ssr-friendly/no-dom-globals-in-module-scope
  typeof window.document !== 'undefined' &&
  // eslint-disable-next-line ssr-friendly/no-dom-globals-in-module-scope
  typeof window.document.createElement !== 'undefined'
    ? useLayoutEffect
    : useEffect

export default useIsomorphicLayoutEffect
