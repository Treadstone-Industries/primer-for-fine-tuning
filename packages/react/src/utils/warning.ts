function emptyFunction() {}

const warn = __DEV__
  ? function warn(message: string) {
      // eslint-disable-next-line no-console
      console.warn('Warning:', message)
    }
  : emptyFunction

// Inspired by warning by fbjs
// @see https://github.com/facebook/fbjs/blob/main/packages/fbjs/src/__forks__/warning.js
const warning = __DEV__
  ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function warning(condition: any, format: string, ...args: Array<any>) {
      if (condition) {
        let index = 0
        const message = format.replace(/%s/g, () => {
          return args[index++]
        })
        warn(message)
      }
    }
  : emptyFunction

export {warn, warning}
