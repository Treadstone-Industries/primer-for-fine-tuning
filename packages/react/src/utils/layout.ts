import theme from '../theme'

// TODO: Update the hard-coded values when the primitives are ready
export const PrimerBreakpoints = {
  xsmall: {
    width: '320px',
  },
  small: {
    width: theme.breakpoints[0],
  },
  medium: {
    width: theme.breakpoints[1],
  },
  large: {
    width: theme.breakpoints[2],
  },
  xlarge: {
    width: theme.breakpoints[3],
  },
  xxlarge: {
    width: '1400px',
  },
}

export function breakpointHigher(width: string): string {
  return `@media screen and (min-width: ${width})`
}

export function breakpointLower(width: string): string {
  return `@media screen and (max-width: calc(${width} - 0.02px))`
}

// Media queries associated with Primer viewport ranges.
export const mediaQueries = {
  narrow: breakpointLower(PrimerBreakpoints.medium.width), // @media screen and (max-width: 768px - 0.02px) // < 768px (max-with is inclusive)
  regular: breakpointHigher(PrimerBreakpoints.medium.width), // @media screen and (min-width: 768) // >= 768px
  wide: breakpointHigher(PrimerBreakpoints.xxlarge.width), // (min-width: 1400px) // >= 1400px
}
