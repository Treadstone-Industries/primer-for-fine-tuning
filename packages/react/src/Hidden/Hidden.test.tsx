import React from 'react'
import {render} from '@testing-library/react'
import {Hidden} from '.'
import MatchMediaMock from 'jest-matchmedia-mock'
import {behavesAsComponent, checkExports, renderStyles, checkStoriesForAxeViolations} from '../utils/testing'
import {mediaQueries} from '../utils/layout'

let matchMedia: MatchMediaMock
describe('Hidden', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock()
  })

  afterAll(() => {
    matchMedia.clear()
  })

  behavesAsComponent({
    Component: Hidden,
    options: {skipAs: true, skipSx: true},
    toRender: () => (
      <Hidden when={'narrow'}>
        <div>Hidden when narrow</div>
      </Hidden>
    ),
  })

  checkExports('Hidden', {
    default: Hidden,
    Hidden,
  })

  it('renders `when` prop as expected', () => {
    const {container} = render(
      <Hidden when={'narrow'}>
        <div>Hidden when narrow</div>
      </Hidden>,
    )
    expect(container).toMatchSnapshot()
  })

  it('renders the styles as expected when a single viewport value is provided as a string via `when` prop', () => {
    const expectedStyles = {
      // `.replace` is used because renderStyles return the JSON object without a space after the column
      [`${mediaQueries.regular.replace(': ', ':')}`]: {
        display: 'none',
      },
    }
    expect(
      renderStyles(
        <Hidden when="regular">
          <div>This is hidden when regular viewports</div>
        </Hidden>,
      ),
    ).toEqual(expect.objectContaining(expectedStyles))
  })

  it('renders the styles as expected when multiple viewport values are provided as an array via `when` prop', () => {
    const expectedStyles = {
      [`${mediaQueries.narrow.replace(': ', ':')}`]: {
        display: 'none',
      },
      [`${mediaQueries.wide.replace(': ', ':')}`]: {
        display: 'none',
      },
    }
    expect(
      renderStyles(
        <Hidden when={['narrow', 'wide']}>
          <div>This is hidden when regular and wide viewports</div>
        </Hidden>,
      ),
    ).toEqual(expect.objectContaining(expectedStyles))
  })
})

checkStoriesForAxeViolations('Hidden.features', '../Hidden/')
checkStoriesForAxeViolations('Hidden.examples', '../Hidden/')
