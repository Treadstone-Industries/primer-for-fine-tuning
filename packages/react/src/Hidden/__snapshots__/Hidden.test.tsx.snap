// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hidden renders \`when\` prop as expected 1`] = `
@media screen and (max-width:calc(768px - 0.02px)) {
  .c0 {
    display: none;
  }
}

<div>
  <div
    class="c0"
  >
    <div>
      Hidden when narrow
    </div>
  </div>
</div>
`;
